### PR TITLE
session: use TxnCtx.InfoSchema, no matter vars.InTxn() or not

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #41622 close #42003

Problem Summary:

### What is changed and how it works?

Build plan need to use the same InfoSchema, but `vars.InTxn()` is false in #41622.
Relax the condition can fix issue/41622

Alternative: change all `sctx.GetInfoSchema()` to `sessiontxn.GetTxnManager(sctx).GetTxnInfoSchema()`
It involves a lot of changes and seems more fragile comparing to this single line of change.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)

Test with the reproduce provided by @crazycs520 in the issue.

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @lcwangchao @crazycs520

File: session/session.go [hunk 1]
```diff
@@ -3982,7 +3982,7 @@ func (s *session) GetInfoSchema() sessionctx.InfoschemaMetaVersion {
 	if snap, ok := vars.SnapshotInfoschema.(infoschema.InfoSchema); ok {
 		logutil.BgLogger().Info("use snapshot schema", zap.Uint64("conn", vars.ConnectionID), zap.Int64("schemaVersion", snap.SchemaMetaVersion()))
 		is = snap
-	} else if vars.TxnCtx != nil && vars.InTxn() {
+	} else if vars.TxnCtx != nil {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply session/session.go:3985]: `vars.InTxn()` is not set when building plan, and the infoschema may change during building (and also during executing)
`vars.TxnCtx.InfoSchema` is set anyway, so relax the `vars.InTxn()` check here can make the building process use the correct infoschema, i.e. `vars.TxnCtx.InfoSchema`

`GetInfoSchema()` spread all over the code base, involves much more change.

**Review by crazycs520 — APPROVED** [review]
LGTM

**Review by lcwangchao — APPROVED** [review]

**Pull request merged**